README.md
# useriq-react-native

* [Installing react-native SDK](https://github.com/useriq-com/useriq-react-native#installing-react-native-sdk)
* [Linking the project](https://github.com/useriq-com/useriq-react-native#linking-the-project)
* [Usage](https://github.com/useriq-com/useriq-react-native#usage)
* [API & Usage](https://github.com/useriq-com/useriq-react-native#api--usage)

---

## Installing React-native SDK

`$ npm install @useriq/useriq-react-native --save`

> Note: Before `npm 5.0.0` the `--save` flag is required in this step. React Native will link modules based on dependencies in your package.json file.

Or in case you prefer yarn over npm, use the following command

`yarn add @useriq/useriq-react-native`

---
## Linking the project
### Autolinking
- **If using React Native 0.60+**


  [CLI autolink feature](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md) links the module while building the app. 


- **If using React Native <= 0.59**


  ```bash
  $ react-native link @useriq/useriq-react-native
  ```

If you can't or don't want to use the CLI tool, you can also manually link the library using the instructions below

### Manual linking
<details>
<summary>Manually link the library on iOS</summary>

1. Install UserIQ framework via Cocoapods in the iOS folder of your app project
  
  ```ruby
  pod 'UserIQ'
  ```
  
2. Follow the [instructions in the React Native documentation](https://facebook.github.io/react-native/docs/linking-libraries-ios#manual-linking) to manually link the framework

</details>

<details>
<summary>Manually link the library on Android</summary>

1. Open up `android/app/src/main/java/[...]/MainApplication.java`

- Add `import com.useriq.rn.UserIQReactNativePackage;` to the imports at the top of the file
- Add `new UserIQReactNativePackage()` to the list returned by the `getPackages()` method

2. Append the following lines to `android/settings.gradle`:
   ```
   include ':@useriq_useriq-react-native'
   project(':@useriq_useriq-react-native').projectDir = new File(rootProject.projectDir, '../node_modules/@useriq/useriq-react-native/android')
   ```
3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
   ```
    implementation project(':@useriq_useriq-react-native')
   ```

</details>

---

## Usage

### 1. Initialization

Initialize `UserIQ` sdk as eary as possible. `UserIQ.init()` method needs to be called atleast once after the root component / application is mounted

```javascript
import React from 'react';
import UserIQ from '@useriq/useriq-react-native'

export class App extends React.Component {

  componentDidMount() {
    UserIQ.init('<YOUR_MOBILE_API_KEY>')
  }
  ...
}
```
> Note: The `<YOUR_MOBILE_API_KEY>` is available in the [mobile dashboard](https://mobile.useriq.com) under the `App Settings` page.

### 1a. Initialize for multiple platforms

If you have single code base for both iOS & Android & if you want to initialize SDK for both platforms, you can initialze using `Platform.select`. If not skip to step 2.

```javascript
import React from 'react';
import { Platform } from 'react-native';
import UserIQ from '@useriq/useriq-react-native'

export class App extends React.Component {

  componentDidMount() {
    Platform.select({
      ios: () => UserIQ.init('IOS_API_KEY'),
      android: () => UserIQ.init('ANDROID_API_KEY'),
    })()
  }
  ...
}
```

This will automatically choose the right `api_key` for the appropriate platform & initialize it. An anonymous user is set and SDK is initiated

> Note: The `IOS_API_KEY` & `ANDROID_API_KEY` is available in the [mobile dashboard](https://mobile.useriq.com) under the `App Settings` page.

### 2. Set loggedin user

SDK initialization itself doesnt send any data to UserIQ server until `setUser()` is called. So after user is sucessfully logged in, call `setUser()` with required params. `id`, `name`, `email`, `accountId`, `accountName` and `signUpDate` are mandatory parameters to be passed and values should be passed `String`

```javascript
import React from 'react'
import UserIQ from '@useriq/useriq-react-native'

class LoginComponent extends React.Component {
  onLoginSuccess(user) {
    UserIQ.setUser({
      id: user.id,
      name: user.name,
      email: user.email,
      accountId: user.accountId,
      accountName: user.accountName,
      signUpDate: user.signUpDate,
      custom_parameter_1: "custom_value_1",
      custom_parameter_2: "custom_value_2"
    })
  }
}
```

### 3. Logout
If a user logs out, the user can be reset to anonymous user just by calling the `logout` API. Make sure this method is called when the user logs out, so that login screen tracking and other information not related to the user does not get linked to the user.

```javascript
  UserIQ.logOut()
```

# API & USAGE

For more details on API & usage, please refer to [wiki page](https://github.com/useriq-com/useriq-react-native/wiki)


CHANGELOG.md
# UserIQ React Native SDK Changelog

All notable changes in Android SDK are documented here. We adhere to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

### [2.6.3] - 2020-01-04

#### Fixed
- Abort installation if cocoapods is not found

---

### [2.6.2] - 2019-12-20

#### Added
- Post-install script to automatically install UserIQ framework
- Updates UserIQ framework to latest version on react-native library update

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#262---2019-12-20)

---
### [2.6.1] - 2019-12-12

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#261---2019-12-12)

### [2.6.0] - 2019-12-04

[Android Changelog](https://github.com/useriq-com/android-sdk/blob/master/CHANGELOG.md#260---2019-12-04)

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#260---2019-12-04)

### [2.5.2] - 2019-11-28

[Android Changelog](https://github.com/useriq-com/android-sdk/blob/master/CHANGELOG.md#252---2019-11-28)

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#252---2019-11-28)

### [2.5.1] - 2019-11-07

#### Updated

[Android Changelog](https://github.com/useriq-com/android-sdk/blob/master/CHANGELOG.md#251---2019-11-07)

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#251---2019-11-07)

---

### [2.5.0] - 2019-10-09

[Android Changelog](https://github.com/useriq-com/android-sdk/blob/master/CHANGELOG.md#250---2019-10-09)

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#250---2019-10-09)

---

### [2.4.2] - 2019-08-30

[Android Changelog](https://github.com/useriq-com/android-sdk/blob/master/CHANGELOG.md#242---2019-08-30)

[iOS Changelog](https://github.com/useriq-com/ios-sdk/blob/master/CHANGELOG.md#242---2019-08-30)

---

### [2.4.1] - 2019-08-08

#### Updated

- Xposed APP_FOREGROUND event to dashboard

---

### [2.4.0] - 2019-07-25

#### Added

- Comment box support in NPS & Star campaigns
- New campaign button action OPEN_WEBVIEW. Opens a desired url in a webview in the app.
- Xamarin automation id support

#### Updated

- Improved screen detection mechanism

#### Fixed

- Added support for custom RootView (for React-native)

---

### [2.3.0] - 2019-07-09

#### Added

- New campaign button action support: `Open Question` & `Start Walkthrough`
- New APIs
  - public static void init(Activity activity, String apiKey)
  - public static void setUser(Context context, User user)
  - public static boolean showCtxHelp(Context context)
  - public static boolean showHelpCentre(Context context)
#### Updated

- Improved UI for helpcenter
- Support for Tooltip styling

#### Deprected

- public static void init(Application application, String apiKey)
- public static void setUser(User user)
- public static boolean showCtxHelp()
- public static boolean showHelpCentre()

---

### [2.2.0] - 2019-06-17

#### Added

- Automatic/Manual tooltip positioning support (2 positions)
- Automatic/Manual beacon and number tooltip positioning support (9 positions)
- Support for `x` and `y` offsets for tooltips
- Ability to make walkthrough wait for the next step for a specific amount of time
- Keyboard support for emulators (on Dashboard)
- Copy/Paste support for emulators (on Dashboard)

#### Updated

- Walkthrough enhancements

#### Fixed

- Minor bug fixes

### [2.1.0] - 2019-05-30

#### Added

- Landscape Orientation support (on Dashboard)

#### Updated

- Added event tracking for campaings (walkthrough & campaigns)

#### Fixed

- Fixed React Native click tracking
- Fixed walkthrough tooltip's button for react native
- Sending App resume event on IO connection
- Other minor bug fixes

### [2.0.4] - 2019-05-11

#### Fixed

- Reuse existing Websocket when setUser is called

#### Added 

- Added debug logs

### [2.0.3] - 2019-04-16

#### Changing user

-- `setUser(User user)` can be used to log in and logout user. Keep a standard user info for logged out users for tracking purposes.

### [2.0.2] - 2019-04-12

#### Minor improvement
- Change in the android API used for getting position of the root view

### [2.0.1] - 2019-04-05

#### React native 
- FAB will be attached as soon as the SDK is initialized.

### [2.0.0] - 2018-12-18

#### Added

- NPS Surveys (3 types)
- Rating Surveys (3 types)
- Walkthroughs (beacons, Numbered & tooltips)
  - Both one off & multiple steps in same screens
  - Cross screen walkthroughs with step triggers
- Helpcenter
  - Rich text support
  - Ability to add images (auto cached on device)
  - Ability to export walkthrough’s into questions
- Contextual Help
  - Configurable guide on per screen basis
  - Allow manual triggers of FAQ’s & Walkthroughs

---


example/e2e/mock-sync-server.js
const WebSocket = require('ws')
const {SimpleRpc} = require('@unfold/simple-rpc')
const {WSTransport} = require('@unfold/simple-rpc')

function start(port, cb) {
  /** @type WebSocket.ServerOptions */
  const opts = {
    path: '/',
    port,
    verifyClient: (info, verifyCb) => verifyCb(true),
  }

  const wss = new WebSocket.Server(opts, cb)
  wss.on('connection', addClient)
  wss.on('error', err => {
    throw err
  })
}


function addClient(ws) {
  const req = ws.upgradeReq
  const id = req.headers['id']
  const kind = req.headers['kind']

  const rpc = new SimpleRpc(new WSTransport(ws),{
    onApp: onApp,
    onCloseApp: onCloseApp,
    onUpdateUser: onUpdateUser,
    onPrefEvert: onPrefEvent,
    onRotation: onRotation,
    captureScreen: onRotation,
    getAppInfo: getAppInfo,
    onEMUEvent: onEMUEvent,
    syncApp: syncApp,
    onLog: onLog,
    onSDKEvent: onSDKEvent,
    getSyncData: getSyncData
  });
}


onApp = async (props, id) => {
  const {userId, apiKey, versionCode, versionString} = props;
}

onCloseApp = async (appProps, id) => {

}

onUpdateUser = async (appProps, userData) => {
  console.log("onUpdateUser called");
}

onPrefEvent = async (p, appProps, id) => {

}

onRotation = async (n, id) => {

}

captureScreen = async (apiKey, id) => {

}

getAppInfo = async (apiKey, id) => {

}

onEMUEvent = async (e, id) => {

}

syncApp = async (data, apiKey, id) => {
  console.log("sync app called");
}

onLog = async (props, e , id) => {

}

onSDKEvent = async (appProps, userData, e, id) => {
  const evts = e['events']
  const first_evt = evts[0]
  const evt_type = first_evt['evt'];
  if (evt_type == 1) {
    return [{"type":"v1Modal","value":"NAVrA4"},{"type":"v1Modal","value":"rAqQAY"},{"type":"v1Modal","value":"O7GZmK"}]
  }
}


getSyncData = async (apiKey) => {
  const syncData = '{"syncVersion":1566308321057,"sdkEnabled":true,"fabEnabled":true,"eventOpts":{"enabled":true,"bufSize":0},"theme":{"color":[255,255,255,255],"textSize":12,"titleColor":[255,255,255,255],"titleSize":16,"bgColor":[230,64,40,79],"borderColor":[255,64,40,79],"borderRadius":4,"borderWidth":1,"fabColor":[255,255,255,255],"fabBgColor":[230,7,136,136],"ctxBtnBorderColor":[96,69,69,69],"ctxBtnColor":[96,69,69,69],"ctxColor":[200,50,50,50],"ctxBgColor":[200,255,255,255]},"logLevel":3,"hcWelcomeText":"We are here to help you with anything and everything related about the app.","trackPerformance":false,"screens":[{"id":"NmE27K","predicate":{"controller":"com.aravind.crawltestapp.FirstViewController","allOf":["ea346500","2e96995e"]},"imageId":"a77249ce-c142-4c98-a2a3-635ab0e90161","elements":[{"id":"NmE27K","key":"ea346500","props":{"accId":"accId"},"bounds":[0,0,750,1334]},{"id":"NAVrA4","key":"2e96995e","props":{"accId":"accId"},"bounds":[296,570,454,630]},{"id":"rAqQAY","key":"294a8efe","props":{"wIndex":0,"text":"Register","cls":"UIButton"},"bounds":[296,704,454,764]}]},{"id":"NAVrA4","predicate":{"controller":"com.aravind.crawltestapp.LoginViewController","allOf":["2313fa9d","365c91a2"]},"imageId":"3bef8f23-63f8-4fe4-9483-d456bae8e0ec","elements":[{"id":"O7GZmK","key":"2313fa9d","props":{"wIndex":0,"text":"Login With Google","cls":"UIButton"},"bounds":[215,561,535,621]},{"id":"57b4l6","key":"365c91a2","props":{"wIndex":0,"text":"Login With Facebook","cls":"UIButton"},"bounds":[215,637,535,697]},{"id":"WAZ2A2","key":"74f2505c","props":{"wIndex":0,"text":"Login With LinkedIn","cls":"UIButton"},"bounds":[215,713,535,773]}]},{"id":"O7GZmK","predicate":{"controller":"com.aravind.crawltestapp.SecondLoggedViewController","allOf":["b2c7e4b6"]},"imageId":"8f76272a-79b9-4317-9c86-82f397c6a9e9","elements":[{"id":"jmN37W","key":"b2c7e4b6","props":{"wIndex":0,"text":"Reload","cls":"UIButton"},"bounds":[328,637,422,697]}]},{"id":"rAqQAY","predicate":{"controller":"com.aravind.crawltestapp.FirstLoggedViewController","allOf":["3fbbc2ff"]},"imageId":"55e57ded-c2b3-4eb1-a9d2-069f10786150","elements":[{"id":"E70NlO","key":"3fbbc2ff","props":{"wIndex":0,"text":"Name : Emp 0","cls":"UILabel"},"bounds":[30,144,245,186]}]},{"id":"57b4l6","predicate":{"controller":"com.aravind.crawltestapp.InfoViewController","allOf":["7b183ad2","09abb12f"]},"imageId":"e29ebd0b-31b8-4e4e-9c6e-50350e909bfb","elements":[{"id":"jmYN7q","key":"7b183ad2","props":{"wIndex":0,"text":"Appunfold Sample App Developed by Aravind G S 2017-2018","cls":"UILabel"},"bounds":[80,536,670,874]},{"id":"27QDmW","key":"09abb12f","props":{"wIndex":0,"text":"Done","cls":"UIButton"},"bounds":[338,890,412,950]}]},{"id":"WAZ2A2","predicate":{"controller":"com.aravind.crawltestapp.RegisterViewController","allOf":["3e190ace","c650359c"]},"imageId":"59fc335b-1616-42a6-8126-95c33be10468","elements":[{"id":"YAekmg","key":"3e190ace","props":{"wIndex":0,"text":"Email","cls":"UITextField"},"bounds":[210,637,540,697]},{"id":"pmRPAY","key":"c650359c","props":{"wIndex":0,"text":"Register","cls":"UIButton"},"bounds":[286,713,464,773]}]},{"id":"E70NlO","predicate":{"controller":"com.aravind.crawltestapp.UIViewController","allOf":["d2e640c2"]},"imageId":"d81d2d8c-3ad9-4c6e-b08c-757f2337015b","elements":[{"id":"GmaQmj","key":"d2e640c2","props":{"wIndex":0,"text":"Continue","cls":"UIButton"},"bounds":[281,728,469,788]}]}],"assets":["c9d07c2a-700c-4d55-b413-68426e156221"],"v1Modal":[{"id":"O7GZmK","name":"Star Default","layout":{"id":"backdrop","type":"node","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0","bgColor":[102,47,47,47],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"children":[{"id":"outsideClick","type":"button","label":"","action":"CLOSE","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0"},{"id":"frame","type":"node","l":"$parent.l+4.1%","t":"$parent.t+23.1%","r":"$parent.r-3.2%","b":"$parent.b-20%","bgColor":[255,255,255,255],"borderRadius":[10,10,10,10],"borderWidth":0,"borderColor":[255,255,255,255],"children":[{"id":"stars","type":"rating","l":"$parent.l+5%","t":"$parent.t+5%","r":"$parent.r-5%","b":"$parent.b-60%","activeColor":[255,248,231,28],"borderRadius":[20,20,20,20],"borderWidth":1,"borderColor":[200,50,50,50],"bgColor":[0,0,0,0],"numStars":5},{"id":"title","type":"text","alignX":"center","alignY":"center","text":"Rate Our Application","textSize":24,"l":"$parent.l+5%","t":"$parent.t+25.7%","r":"$parent.r-5%","b":"$parent.b-53.1%","color":[255,50,50,50],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[0,0,0,0]},{"id":"content","type":"text","alignX":"center","alignY":"center","text":"How would you rate this feature?","textSize":20,"l":"$parent.l+5%","t":"$parent.t+35.3%","r":"$parent.r-5%","b":"$parent.b-30.7%","color":[255,50,50,50],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[0,0,0,0]},{"id":"close","type":"button","label":"SUBMIT","action":"SUBMIT","l":"$parent.l+27.3%","t":"$parent.t+79.4%","r":"$parent.r-27.6%","b":"$parent.b-8.3%","textSize":15,"color":[200,255,255,255],"bgColor":[255,16,98,162],"borderRadius":[30,30,30,30],"borderWidth":0,"borderColor":[255,255,255,255]},{"id":"cross","type":"button","label":"×","action":"CLOSE","l":"$parent.l+90%","t":"$parent.t","r":"$parent.r","b":"$parent.t+10%","textSize":32,"color":[255,0,0,0],"bgColor":[0,0,0,0],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,245,195,0]},{"id":"editText","type":"editText","l":"$parent.l+5%","t":"$parent.t+65%","r":"$parent.r-5%","b":"$parent.t+75%","borderRadius":[24,24,24,24],"borderWidth":1,"borderColor":[155,155,155,155],"bgColor":[255,255,255,255],"placeholder":"Please tell us why you gave this rating"}]}]},"assets":[],"type":"starDefault"},{"id":"rAqQAY","name":"NPS Default","layout":{"id":"backdrop","type":"node","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0","bgColor":[102,47,47,47],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"children":[{"id":"outsideClick","type":"button","label":"","action":"CLOSE","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0"},{"id":"frame","type":"node","l":"$parent.l+7.3%","t":"$parent.t+25.7%","r":"$parent.r-7.1%","b":"$parent.b-18.2%","bgColor":[255,255,255,255],"borderRadius":[10,10,10,10],"borderWidth":0,"borderColor":[255,255,255,255],"children":[{"id":"title","type":"text","alignX":"center","alignY":"center","text":"How likely are you to recommend our product to someone?","textSize":15,"l":"$parent.l+0%","t":"$parent.t+0%","r":"$parent.r+0%","b":"$parent.t+30%","color":[255,255,255,255],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,16,98,162]},{"id":"cross","type":"button","label":"×","action":"CLOSE","l":"$parent.l+92%","t":"$parent.t","r":"$parent.r","b":"$parent.t+8%","textSize":22,"color":[255,255,255,255],"bgColor":[0,0,0,0],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,245,195,0]},{"id":"npsFrame","type":"nps","l":"$parent.l","t":"$parent.t+30%","r":"$parent.r","b":"$parent.t+70%","borderRadius":[0,0,0,0],"borderWidth":0,"activeColor":[255,216,103,78],"bgColor":[255,255,255,255],"numStart":0,"numEnd":10},{"id":"label1","type":"text","alignX":"left","alignY":"center","text":"0-Not likely","textSize":11,"l":"$parent.l+5%","t":"$parent.t+70%","r":"$parent.l+45%","b":"$parent.t+75%","color":[255,0,0,0],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,255,255,255]},{"id":"label2","type":"text","alignX":"right","alignY":"center","text":"10-Extremely likely","textSize":11,"l":"$parent.l+55%","t":"$parent.t+70%","r":"$parent.r-5%","b":"$parent.t+75%","color":[255,0,0,0],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,255,255,255]},{"id":"close","type":"button","label":"SUBMIT","action":"SUBMIT","l":"$parent.l+27.3%","t":"$parent.t+82%","r":"$parent.r-27.6%","b":"$parent.t+93%","textSize":15,"color":[255,255,255,255],"bgColor":[255,16,98,162],"borderRadius":[30,30,30,30],"borderWidth":0,"borderColor":[255,255,255,255]}]}]},"assets":[],"type":"npsDefault"},{"id":"NAVrA4","name":"Sample Modal","layout":{"id":"backdrop","type":"node","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0","bgColor":[102,47,47,47],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"children":[{"id":"outsideClick","type":"button","label":"","action":"CLOSE","l":"$parent.l+0","t":"$parent.t+0","r":"$parent.r+0","b":"$parent.b+0"},{"id":"frame","type":"node","l":"$parent.l+12%","t":"$parent.t+17.5%","r":"$parent.r-12%","b":"$parent.t+82.5%","bgColor":[255,255,255,255],"borderRadius":[10,10,10,10],"borderWidth":1,"borderColor":[255,255,255,255],"children":[{"id":"Image","type":"image","assetId":"c9d07c2a-700c-4d55-b413-68426e156221","crop":true,"alignX":"center","alignY":"center","l":"$parent.l+0","t":"$parent.t","r":"$parent.r+0","b":"$parent.t+40%","bgColor":[255,255,255,255],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255]},{"id":"title","type":"text","alignX":"center","alignY":"center","text":"Your Headlines Goes Here","textSize":22,"l":"$parent.l+5%","t":"$parent.t+45%","r":"$parent.r-5%","b":"$parent.t+60%","color":[255,54,172,173],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,255,255,255]},{"id":"body","type":"text","alignX":"center","alignY":"center","text":"Tell your users more but not too much. We recommend 50 words or less","textSize":13,"l":"$parent.l+5%","t":"$parent.t+60%","r":"$parent.r-5%","b":"$parent.t+75%","color":[255,75,75,75],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,255,255,255]},{"id":"footer","type":"node","l":"$parent.l","t":"$parent.t+75%","r":"$parent.r","b":"$parent.b","borderRadius":[0,0,20,20],"borderWidth":0,"borderColor":[255,255,255,255],"bgColor":[255,255,255,255],"children":[{"id":"button1","type":"button","label":"LET\'S GO","action":"CLOSE","l":"$parent.l+15.5%","t":"$parent.t+5%","r":"$parent.r-15.5%","b":"$parent.t+40%","textSize":14,"color":[255,255,255,255],"bgColor":[255,54,172,173],"borderRadius":[30,30,30,30],"borderWidth":0,"borderColor":[255,255,255,255]},{"id":"button2","type":"button","label":"LATER","action":"CLOSE","l":"$parent.l+15.5%","t":"$parent.t+50%","r":"$parent.r-15.5%","b":"$parent.t+85%","textSize":14,"color":[255,54,172,173],"bgColor":[255,255,255,255],"borderRadius":[30,30,30,30],"borderWidth":3,"borderColor":[255,54,172,173]}]},{"id":"cross","type":"button","label":"×","action":"CLOSE","l":"$parent.l+92%","t":"$parent.t-1%","r":"$parent.r-1%","b":"$parent.t+10%","textSize":25,"color":[255,85,85,85],"bgColor":[0,0,0,0],"borderRadius":[0,0,0,0],"borderWidth":0,"borderColor":[255,245,195,0]}]}]},"assets":["c9d07c2a-700c-4d55-b413-68426e156221"],"type":"inAppDefault"}],"qGroup":[{"id":"NmE27K","name":"Group 1","questions":[{"id":"NmE27K","name":"Question 1","desc":[{"type":"rtf","value":"<p>Answer 1</p>"}]}]}],"walkthrough":[],"ctxHelp":[]}';
  return syncData
}


/**
 * start listens to connections from Real devices
 */
start(4444, () => {
  console.log('Mock Sync Server started at ' + 4444)
})

